### ISSUE TITLE
Allow for TagsAttribute to be inherited from interfaces

### ISSUE THREAD

**Issue by zendever** [fluentmigrator/fluentmigrator#1026]
This is an extension of older tickets submitted #626 and #648.

I found a use case for allowing tags to be "combined" by various classifications. So, for example, we have environment specific tags ("Production", "Non-Production", "QA", etc.) but we also have database state tags ("New") which only be allowed to run on a new database instance, or "Testing" which might only seed data for testing.

Because both classifications can exist simultaneously, I found a suitable solution would be to allow tags to be inherited from multiple interfaces as well. So, for a case where we might want to do QA-specific test data when working on standing up automated integration tests. Since we have 2 different types of QA-level testing, it's possible that a QA automated test would have specific information that something like a exploratory QA test environment wouldn't.

I will submit a pull request for this change.

**Comment by jzabroski** [comment]
For testing, we use maintenance stage AfterAll and a `UnitTest` tag.  Actual seeding migrations (e.g., for production fixes such as updating federally mandated rates on a certain effective date) are tagged with `[Tags(TagName.Development, TagName.Alpha, TagName.Beta, TagName.Production)]` - effectively everything but `TagName.UnitTest`

I also recommend using an object hydration library like AutoFixture, ObjectHydrator, etc. to create sample data.  I prefer AutoFixture for testing, and tools like ObjectHydrator for populating QA / SalesDemo databases since it's an easy way to add some humor to the QA environment and Sales demoes.

**Comment by jzabroski** [comment]
@zendever How do you want this to work, exactly?

Taking the example from the documentation (which should be updated if we change how this works)...

```csharp
[Tags("DK", "NL", "UK")]
[Tags("Staging", "Production")]
[Migration(1)]
public class DoSomeStuffToEuropeanStagingAndProdDbs : Migration { /* ... etc ... */ }
```

Would you refactor this into the following:

```csharp
[Tags("DK", "NL", "UK")]
public interface ITagsCountryDkNlUk { } // marker interface for applying tags
[Tags("Staging", "Production")]
public interface ITagsStagingOrProduction { }
[Migration(1)]
public class DoSomeStuffToEuropeanStagingAndProdDbs : 
    ITagsCountryDkNlUk,
    ITagsStagingOrProduction,
    Migration
{
    /* ... etc ... */
}
```

Or, perhaps you mean something like this, which is already supported and far more discoverable since I would not want to personally maintain code where the behavior is dynamically determined by whether or not an interface is implemented - I think Attributes are the right way to determine this stuff:

1.  
```csharp
public class TagsStagingOrProduction : TagsAttribute
{
        public TagsAttribute(TagBehavior behavior)
        {
            Behavior = behavior;
            TagNames = new string[] { "Staging", "Production" };
        }
}
```
2. Note, `TagsAttribute` already allows inherited attributes (`Inherited` value on `AttributeUsage` defaults to `true`):
```csharp
    [AttributeUsage(AttributeTargets.Class, AllowMultiple = true)]
    public class TagsAttribute : Attribute
```

**Comment by zendever** [comment]
I would vote for the first example whereby the apply to the interface.

The interface implementation only allows for the combination of multiple tag sets. Hypothetically, I could foresee having state-regulatory shards that might be combined at a timezone-level or any number of ways, honestly. Rather than having to run the migration commands to differ for every permutation of tag sets, the migrations may simply be by state, and the migration, itself, would be something that might apply to one more combinations of those factors.

So, let's take the shards by state...if I have data that might be applicable to more than one timezone, I might create an interface that takes tags for a specific timezone (IApplyToCentralTime, IApplyToEasternTime, etc.). While I certainly COULD add tags for each timezone rather than creating an interface, I'd also have to run multiple versions of a shard's migration if I want to implement run those migrations tagged with both a specific state requirement AND a requirement that applies to something in a specific timezone. This would mean that I'd have to run both "GA" and "Eastern Time" as separate migrations even though I know that the "key" is really "GA", but may apply to more than one state. It also means that there's really no way to run applicable migrations across these tag subsets in order. It gets increasingly more complex if you have something like an implementation type (IApplyToOfflineFTPConsumer, IApplyToOnlineRESTConsumer, etc.) that might still be applicable at a shard level. So if my TX shard implements both IApplyToCentralTime and IApplyToOnlineRESTConsumer, without composite tags via multiple interface inheritance, I'd have to run all TX-specific tags, then time-zone tags, and then consumer-type tags. However, if I keep tags associated by shard, then I wouldn't necessarily have to copy/paste all of the associated tags for an applicable migration NOR would have to worry about creating multiple migrations by tag set...I would just ensure that the targeted shard was in one of the interfaces (if necessary).

I know this may flip tags on their heads a bit, but at least allowing for the use of tags in this way allows for migrations either by tag segregation (e.g. by environment) or migrations by relevant features (i.e. timezone, regulatory jurisdiction, implementation requirements, etc.).

**Comment by jzabroski** [comment]
It sounds like what you want are xunit.net-style tags, called Traits (see: http://www.brendanconnolly.net/organizing-tests-with-xunit-traits/). Traits are basically a Tag plus a Tag Context.

I don't think you want marker interfaces, because of the discoverability requirements. If you want to do this yourself, you can implement your own runner, but even then my recommendation is to put more structure around it, like having all tag marker interfaces inherit a God Marker Interface `ITag`. And if you find you can't write a custom runner that does this tag filtering, submit a PR to make that aspect more pluggable and testable. - But if you want to enhance the existing experience, such as adding an ITag marker interface, let's discuss whether that belongs to the core library. 

I'm also just trying to save you some trouble by helping you discover your true requirements. It would be helpful if you wrote user stories in user-action-benefit triples.

Consider the following three refinements:
1. “As a Spotify user, I would like to add songs to a playlist.” Can I break this down further? (Yes).
2. “As a Spotify user, I would like a menu option to add a song to a playlist.” Specifying the mechanism.
3. “As a Spotify user, I would like new songs to be added to the bottom of the playlist.” Specifying the policy enacted by the mechanism.

In these examples, the benefits aren't super clear. The user and action are super clear in 2 and 3, but in different ways. Articulating the benefits may not be necessary, but it can help tackle complexity when users ask for contradictory requirements:

4. “As a Spotify user, I would like new songs to be added to the **top**  of the playlist.” Specifying the policy enacted by the mechanism; conflicts with story 3. Can it be enhanced by specifying what the benefits are? (Yes.)
5. “As a Spotify user, I would like new songs to be added to the **top**  of the Playlist, so that when I'm talking to friends on Facebook Messenger and they mention a song I should listen to, I can immediately listen to it without losing my playlist.” (Ugh, that's so annoying, Spotify/Facebook!! :) ) 

In your examples, I'm trying to see the essence of what you want.

**Comment by zendever** [comment]
Okay, you're asking me for specific requirements, and I'll be honest, I just want to inherit tags from multiple interfaces. Why? Because it seemed easy and somewhat declarative, as well as discover-able. So, let me build a scenario.

Let's say I run a Healthcare Statistics service for state agencies whereby hospitals need to interact with those state agencies. Because our state agencies request that our data be segregated and audited at the state level, the company has decided to create 1 database for each state (so, 50 databases). The company has several different applications available to the state provider as well as a few different options for processing. We also, periodically need to run data fixes that apply time-sensitive changes and may affect, retroactively, data based on geographical location.

The one thing our CI/CD pipeline knows for certain is that our databases are segregated by state, so for our migrations, we'd simply like to apply the state tag, and all migrations that affect that state's database will be applied.

`c#
[Tags("GA", "CA", "FL", "TX")]
public interface IApplyToReportType1Service { }

[Tags("AL", "AR", "TN")]
public interface IApplyToReportType2Service { }

[Tags("NV", "MI")]
public interface IApplyToReportType3Service { }

[Tags("TX", "AL")]
public interface IApplyToOfflineService { }

[Tags("MI", "FL", "NV")]
public interface IApplyToSynchronousService { }

[Tags("CA", "GA")]
public interface IApplyToAsynchronousService { }

[Tags("GA", "FL", "AL", "TN")
public interface IApplyToEasternTimezone { }

[Tags("CA", "NV")
public interface IApplyToPacificTimezone { }

[Tags("TX", "AR", "MI")
public interface IApplyToCentralTimezone { }

public interface IApplyToEasternUS : IApplyToEasternTimezone, IApplyToCentralTimezone { }

// use of tags
[Migration(1)]
public class V1_ApplyToAll : Migration { ... }

[Migration(2)]
public class V2_ApplyToOnlineCustomers : Migration, IApplyToSynchronousService, IApplyToAsynchronousService { ... }

[Migration(3)]
public class V3_ApplyToOfflineCustomers : Migration, IApplyToOfflineService { ... }

[Migration(4)]
public class V4_ApplyToReportType1Customers : Migration, IApplyToReportType1Service { ... }

[Migration(5)]
public class V5_ApplyToAllAgain : Migration { ... }

[Migration(6)]
public class V6_ApplyToReportTypes1And3Customers : Migration, IApplyToReportType1Service, IApplyToReportType3Service { ... }

[Migration(7)]
public class V7_ApplyToImplementationsThatCrossBoundaries : Migration, IApplyToReportType1Service, IApplyToOfflineService { ... }

// etc, etc.
`

Yes, there's many ways I can skin this cat, but IF I wanted to add the state of "NY", all I'd have to do is append "NY" to the appropriate interface, and set up my CI/CD pipeline like every other state implementation. The other options include modifying every applicable migration OR break-up migration runs between tag classifications. Neither of those seem as clean as allowing multiple interface inheritance.

Yes, you could argue the semantics of tag vs. trait, but at the end of the day, all of the examples we've seen for the use of Tag still is, effectively, a trait...the examples you gave included a country trait and an environment trait. Furthermore, allowing for tag inheritance by single inheritance doesn't seem to impact tag discovery any more than multiple inheritance would.

**Comment by jzabroski** [comment]
I see your point.  In languages that support traits directly (like Scala), traits _are_ interfaces, so the distinction I am making is a bit ridiculous.  I'm effectively penalizing you for the fact C# (and the CIL) doesn't support traits.  I'm sorry.

I also see that your requirement has nothing to do with non-functional requirements (sharding for performance) but more regulatory hold requirements.  I did not consider that.

Just curious, what do you do when you have more than 4 states that are applicable, given TagsAttribute stops at 4 tags?

To understand where I am coming from, what you're asking for is not standard .NET behavior.  Even if you just change :
```csharp
[AttributeUsage(AttributeTargets.Class, AllowMultiple = true)]
public class TagsAttribute : Attribute
```
to:
```csharp
[AttributeUsage(AttributeTargets.Class | AttributeTargets.Interface, AllowMultiple = true)]
public class TagsAttribute : Attribute
```
Nothing will happen, because classes don't inherit attributes from their interfaces.  We'd have to change https://github.com/fluentmigrator/fluentmigrator/blob/e82aafa20e6dbe3cefa221303fe23cf8bf59fffd/src/FluentMigrator.Runner.Core/Infrastructure/DefaultMigrationConventions.cs#L95

This goes back to my suggestion that the more basic solution here would be to sub-class DefaultMigrationRunnerConventions and override https://github.com/fluentmigrator/fluentmigrator/blob/e82aafa20e6dbe3cefa221303fe23cf8bf59fffd/src/FluentMigrator.Runner.Core/Infrastructure/DefaultMigrationConventions.cs#L48
and 
https://github.com/fluentmigrator/fluentmigrator/blob/e82aafa20e6dbe3cefa221303fe23cf8bf59fffd/src/FluentMigrator.Runner.Core/Infrastructure/DefaultMigrationConventions.cs#L49

**Comment by zendever** [comment]
So, I thought I saw that tag names was a parameter array of strings...but since it allows multiples, I could just combine...not that I'd want to.

As far as the combined interface, I already have code that works to implement this change and I could submit a pull request.

**Comment by jzabroski** [comment]
Hmm, yeah, that's true.

@fubar-coder  Why do we have those explicit tag overloads on TagsAttribute?  I honestly don't know which C# would select given `[Tags("A", "B", "C", "D")]` - is there some kind of backward compatibility requirement, or support for using named parameters we want to keep?  I think named parameters might be the only way to explicitly select the four parameter attribute constructor.  C# brain teaser 🤧

**Comment by fubar-coder** [comment]
@jzabroski The reason for multiple `TagsAttribute` constructors is (most likely) that using arrays in attributes isn't CLS compliant (IIRC).

**Comment by zendever** [comment]
...that said, I don't think this affects my specific implementation requirements, AFAIK. Just to make sure I wasn't hallucinating, I did set 20 tags in a single Tag Attribute.

**Comment by jzabroski** [comment]
Yeah, I had never paid attention to the fact there is that params overload.  I have only needed 4 tags, to be honest.  I have 5 tags total, but only 4 (or implicitly all) apply at any given time: Development, Alpha, Beta, Production. The fifth is UnitTest but if all five apply, I just don't apply any tags.

By the way, to round out my thoughts on why I was being finicky asking about requirements: It's because we have features like Profiles and Maintenance Migrations and I wonder what the best overall API is given all these different ways to configure migrations.  While it makes some sense to just take the PR and :shipit: , the average newcomer might get confused if there is no guidance and when to use what approach.  Plus, while it might work well either way, development teams might waste time debating it.

**Comment by zendever** [comment]
I understand, which is why I wanted to have this discussion. I do think we have organically discovered a use case that justifies the implementation...unless I was somehow mistaken.

I also took a look at both profiles and maintenance migrations. I do have a question, because it wasn't quite clear...but are profile migrations run only once or run every time? As far as maintenance migrations, I do think the guidance on them is explanation enough...I dig it.

I'll code up the change and send the pull request.

**Comment by jzabroski** [comment]
The only state in FluentMigrator are migrations.  You need a Version Id in order to have state across runs.

I stopped using Profiles for anything, because Tags+`[Maintenance(MigrationStage.AfterAll)]` solves all my problems, personally.  I still have some projects on 1.6.2 using Profiles.  So far on 3.x, I have not needed Profiles.

**Comment by jzabroski** [comment]
> I do think we have organically discovered a use case that justifies the implementation...unless I was somehow mistaken.

Yes, I agree.  But Profiles are actually a really good example of how something got introduced and then we're wondering why we still have it.  The main use case I can think of is seeding/idempotent scripts.  But for some seeding/idempotent script scenarios, the solution doesn't work.  For example, synonyms.  I need synonyms to exist BeforeAll.  You can think of creating synonyms as a sort of seeing for code objects, instead of data objects.

**Comment by fubar-coder** [comment]
Fixed by PR #1027
